cfpb/cfgov-refresh
# Refactor data research API

The initial API was exploratory and leaned on a metadata helper class
called FIPS. Now that all data structures are better understood, we can
refactor the API to remove reliance on the FIPS helper, simplify
save and aggregation methods and speed up operations by normalizing metadata.

## Additions
- data model for "non-MSA" areas, which represent counties in a state that are not part of any Metropolitan Statistical Area. Some may call it rural; I call it antimatter.
- new geo models for county, metro area and state to handle metadata.

The refactor implements excellent suggestions from @chosak and @richaagarwal to remove reliance on the FIPS class, simplify save and data aggregation methods and make use of the [django-jsonfield](https://pypi.python.org/pypi/jsonfield) module.

## Testing
To see the API in action:
- Run migrations
  ```
  python cfgov/manage.py migrate --noinput
  ```
- Download the data set http://files.consumerfinance.gov.s3.amazonaws.com/data/mortgage-performance/initial_data/mortgagedata_refactor2.sql.gz
- Load it
  ```
  gunzip -c ~/Downloads/mortgagedata_refactor2.sql.gz | mysql -uroot v1
  ```
Check out some endpoints:
    - [county time series data](http://localhost:8000/data-research/mortgages/api/v1/time-series/90/12081/)
    - [metro map data](http://localhost:8000/data-research/mortgages/api/v1/map-data/90/metros/2009-01/)
    - [metadata: sampling dates](http://localhost:8000/data-research/mortgages/api/v1/metadata/sampling_dates/)
